src/tty/mod.rs

```diff
@@ -53,6 +53,29 @@ pub trait EventedReadWrite {
     fn write_token(&self) -> mio::Token;
 }
 
+/// Events concerning TTY child processes
+#[derive(PartialEq)]
+pub enum ChildEvent {
+    /// Indicates the child has exited
+    Exited
+}
+
+/// A pseudoterminal (or PTY)
+///
+/// This is a refinement of EventedReadWrite that also provides a channel through which we can be
+/// notified if the PTY child process does something we care about (other than writing to the TTY).
+/// In particular, this allows for race-free child exit notification on UNIX (cf. `SIGCHLD`).
+pub trait EventedPty : EventedReadWrite {
+    #[cfg(unix)]
+    fn child_event_token(&self) -> mio::Token;
+
+    /// Tries to retrieve an event
+    ///
+    /// Returns `Some(event)` on success, or `None` if there are no events to retrieve.
+    #[cfg(unix)]
+    fn next_child_event(&mut self) -> Option<ChildEvent>;
```

Comment:
We already have a `next_child_event` method, so that would likely raise the question why `EventedPty` doesn't just have a `next` function which returns an `Option<ChildEvent>` using the Iterator trait. And I think that might actually be the correct thing to do here, since a `next_*` method which acts pretty much exactly like an iterator but isn't implemented as an iterator is a bit odd idiomatically.

Just changing this to `pub trait EventedPty: EventedReadWrite + Iterator<Item = ChildEvent>` should take care of the `next` method automatically. The obvious drawback here would be that this is more difficult to do cleanly with the platform-specific annotations, but I don't think that's a huge deal.